tests/test_tools.py

```diff
@@ -14,11 +14,128 @@ def test_angular_avg_nd_3():
         < 0.05
     )
+    assert (
+        np.max(
+            np.abs(
+                (p_k_lin[6:, 0] - k_av_bins_lin[6:] ** -2.0) / k_av_bins_lin[6:] ** -2.0
+            )
+        )
+        < 0.05
+    )
+
+
+def test_interp_w_weights():
+    x = np.linspace(-3, 3, 40)
+    P = np.ones((40, 40, 40))
+    P[:4, 3:6, 7:10] = 0
+    weights = np.ones_like(P)
+    weights[:4, :, :] = 0
+    weights[:, 3:6, :] = 0
+    weights[:, :, 7:10] = 0
+
+    # Test 4D avg works
+    freq = [x, x, x]
+    p_k_lin, k_av_bins_lin = angular_average(
+        P,
+        freq,
+        bins=10,
+        interpolation_method="linear",
+        weights=weights,
+        angular_resolution=0.1,
+    )
+
+    assert np.all(p_k_lin == 1.0)
+
+
+def test_angular_avg_nd():
+    x = np.linspace(-3, 3, 40)
+    X, Y, Z = np.meshgrid(x, x, x)
+    r2 = X**2 + Y**2 + Z**2
+    P = r2**-1.0
+
+    # Test 4D avg works
+    P = np.repeat(P, 10).reshape(40, 40, 40, 10)
+    freq = [x, x, x, np.linspace(-2, 2, 10)]
+    p_k_lin, k_av_bins_lin = angular_average(
+        P, freq, bins=10, interpolation_method="linear", weights=np.ones_like(P)
+    )
+
+    p_k_lin, k_av_bins_lin = angular_average_nd(
+        P, freq, bins=10, n=3, interpolation_method="linear"
+    )
+    assert (
+        np.max(
+            np.abs(
+                (p_k_lin[6:, 0] - k_av_bins_lin[6:] ** -2.0) / k_av_bins_lin[6:] ** -2.0
+            )
+        )
+        < 0.05
+    )
+
+    p_k_lin, k_av_bins_lin = angular_average_nd(
+        P, freq, bins=10, n=2, interpolation_method="linear"
+    )
+    assert (
+        np.max(
+            np.abs(
+                (p_k_lin[6:, len(x) // 2, 0] - k_av_bins_lin[6:] ** -2.0)
+                / k_av_bins_lin[6:] ** -2.0
+            )
+        )
+        < 0.05
+    )
+
+    p_k_lin, k_av_bins_lin = angular_average_nd(
+        P, freq, bins=10, n=1, interpolation_method="linear"
+    )
+    assert (
+        np.max(
+            np.abs(
+                (p_k_lin[6:, len(x) // 2, len(x) // 2, 0] - k_av_bins_lin[6:] ** -2.0)
+                / k_av_bins_lin[6:] ** -2.0
+            )
+        )
+        < 0.05
+    )
+
+
+def test_angular_avg_nd_complex_interp():
+    x = np.linspace(-3, 3, 400)
+    X, Y = np.meshgrid(x, x)
+    r2 = X**2 + Y**2
+    P = r2**-1.0 + 1j * r2**-1.0
+    P = np.repeat(P, 100).reshape(400, 400, 100)
+    freq = [x, x, np.linspace(-2, 2, 100)]
+    p_k_lin, k_av_bins_lin = angular_average_nd(
+        P, freq, bins=50, n=2, interpolation_method="linear"
+    )
+    real = np.real(p_k_lin)
+    imag = np.imag(p_k_lin)
+    assert (
+        np.max(
+            np.abs(
+                (real[6:, 0] - k_av_bins_lin[6:] ** -2.0) / k_av_bins_lin[6:] ** -2.0
+            )
+        )
+        < 0.05
+    )
+
+    assert (
+        np.max(
+            np.abs(
+                (imag[6:, 0] - k_av_bins_lin[6:] ** -2.0) / k_av_bins_lin[6:] ** -2.0
+            )
+        )
+        < 0.05
+    )
```

Comment:
maybe an easier test here is just to test that the real and imaginary parts of the result are equal